### ISSUE TITLE
proposal: unit tests in browser

### ISSUE THREAD

**Issue by DanielRuf** [snipsco/yett#2]
We can use https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagesetrequestinterceptionvalue to check if the requests are really blocked.

Example implementation: https://github.com/DanielRuf/gdpr-checks/blob/master/index.js#L18

Might be useful for automatic test using Travis CI to check if everything is blocked as expected.

**Comment by elbywan** [comment]
Hi @DanielRuf !

Thanks for the proposal, it would indeed be very useful to have some tests to check whether the requests are blocked or not and to ensure that the execution is prevented.

The issue with using `Pupeteer` is that it is Chrome based. We would really like to implement a solution that covers the four major browsers, like [`Browserstack`](https://www.browserstack.com/).

It would actually be quite simple to do using [`Jasmine`](https://jasmine.github.io/) unit tests:

- For the loading part, check whether the script was requested using the performance API 
```js
performance.getEntries().find(_ => _.initiatorType === 'script' && _.name === SCRIPT_URL)
```
- For the execution part, make loaded scripts define a global variable and check whether this global variable is `undefined`.

Would you be interested in implementing that kind of unit tests if we take care of the Browserstack setup?

**Comment by DanielRuf** [comment]
> Would you be interested in implementing these kind of unit tests if we take care of the Browserstack setup?

I will try to save some time for it and make some plans for the needed setup and tests. Let's see how we can bring this forward together.

**Comment by elbywan** [comment]
> I will try to save some time for it and make some plans for the needed setup and tests. Let's see how we can bring this forward together.

Great! 👍

**Comment by elbywan** [comment]
@DanielRuf Sorry for the delay, but I just created PR https://github.com/snipsco/yett/pull/8 that adds CI browser testing. Better late than never 😉.

I actually used a `karma.js + mocha + chai` stack instead of `Browserstack + jasmine` as it was much simpler to set up, testing on `Chrome` `Firefox` and `Safari`.

Unit tests also cover most use cases and ensure that scripts are properly blacklisted /whitelisted, and that script having the `javascript/blocked` attributes are not preloaded for `Chrome` and `Safari`.
